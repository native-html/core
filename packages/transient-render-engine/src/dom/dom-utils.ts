import { Text, Element, Node, Document, NodeWithChildren } from 'domhandler';
import { Text as TextType, Tag as TagType } from 'domelementtype';

export { Text, Element, Node, NodeWithChildren, Document };

export function isDomText(node: any): node is Text {
  return node && node.type === TextType;
}

export function isDomElement(node: any): node is Element {
  return node && node.type === TagType;
}
