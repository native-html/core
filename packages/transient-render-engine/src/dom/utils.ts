import { Text as TextType, Tag as TagType } from 'domelementtype';
import { Node, Text, Element } from 'domhandler';

export function isTextNode(node: Node | null | undefined): node is Text {
  return !!node && node.type === TextType;
}

export function isElementNode(node: Node | null | undefined): node is Element {
  return !!node && node.type === TagType;
}
