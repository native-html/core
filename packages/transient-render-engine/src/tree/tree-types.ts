import { Element, Text } from '../dom/dom-utils';
import HTMLContentModel from '../model/HTMLContentModel';
import HTMLElementModel from '../model/HTMLElementModel';
import { TStylesShape } from '../styles/TStyles';
import { TStylesMerger } from '../styles/TStylesMerger';

export interface TNodeInit {
  readonly elementModel: HTMLElementModel<string, HTMLContentModel> | null;
  readonly context: {
    readonly stylesMerger: TStylesMerger;
    readonly setMarkersForTNode?: SetMarkersForTNode;
    readonly removeLineBreaksAroundEastAsianDiscardSet: boolean;
  };
  readonly parent?: TNodeImpl | null;
  readonly domNode?: Element | null;
  readonly styles?: TStylesShape | null;
  readonly parentStyles?: TStylesShape | null;
  readonly nodeIndex?: number;
  readonly isUnregistered?: boolean;
}

/**
 * A minimal TNode shape accessible at instantiation.
 *
 * @public
 */
export interface TNodeDescriptor {
  tagName: string | null;
  classes: string[];
  id: string | null;
  attributes: Record<string, string>;
  markers: Markers;
}

/**
 * A function to set markers for a specific {@link TNode}.
 *
 * @param targetMarkers - The markers instance to alter.
 * @param parentMarkers - Th direct ascendant markers instance for this tnode.
 * @param tnode - The minimalistic {@link TNodeDescriptor} to which the
 * `targetMarkers` will be attached.
 *
 * @public
 */
export type SetMarkersForTNode = (
  targetMarkers: Markers,
  parentMarkers: Readonly<Markers>,
  tnode: TNodeDescriptor
) => void;

/**
 * Markers form an abstraction in which one node provides semantic information
 * to itself and all its descendants. For example, `ins` elements, which stand
 * for "insertion" of content in the context of an edit will provide the {
 * edits: 'ins' } marker to all its descendants.
 *
 * @remarks
 * Custom renderers can use markers to change their layout and convey their
 * semantic meaning. Markers can be derived from attributes, such as `lang` and
 * `dir` attributes, or tag names, such as `a`, `ins`, `del`...
 *
 * **TypeScript users**: You can add fields to the {@link Markers} interface via
 * {@link https://www.typescriptlang.org/docs/handbook/declaration-merging.html#module-augmentation | module augmentation}:
 *
 * ```ts
 * declare module '@native-html/transient-render-engine' {
 *   interface Markers {
 *     customField?: boolean;
 *   }
 * }
 * ```
 */
export interface Markers {
  /**
   * If this node is an `a` or has one as ancestor, this field will be set to
   * `true`.
   *
   * @defaultValue false
   */
  anchor: boolean;
  /**
   * If this node is a `del` or `ins` or has either as ancestor, this field
   * will be set accordingly. Otherwise, it will be set to 'none'.
   *
   * https://html.spec.whatwg.org/#edits
   *
   * @defaultValue 'none'
   */
  edits: 'ins' | 'del' | 'none';
  /**
   * The direction for this content. Follows `dir` attribute.
   *
   * https://html.spec.whatwg.org/#the-dir-attribute
   *
   * @defaultValue 'ltr'
   */
  direction: 'ltr' | 'rtl';
  /**
   * The language for this content. Follows `lang` attribute.
   *
   * https://html.spec.whatwg.org/#the-lang-and-xml:lang-attributes
   */
  lang: string;
  /**
   * * -1: this node is not an `ol` and has no `ol` parents
   * * 0: this node is an `ol` or has one `ol` parent
   * * 1: ...
   */
  olNestLevel: number;
  /**
   * * -1: this node is not an `ul` and has no `ul` parents
   * * 0: this node is an `ul` or has one `ul` parent
   * * 1: ...
   */
  ulNestLevel: number;
  /**
   * Create a new marker instance which extends this markers.
   */
  extend(this: Markers): Markers;
  /**
   * Return a string represenation of this marker, including inherited
   * properties from parent markers.
   */
  toString(): string;
}

/**
 * @public
 */
export interface TNodePrintOptions {
  /**
   * Include styles in snapshot.
   */
  withStyles: boolean;
  /**
   * Include node index in snapshot.
   */
  withNodeIndex: boolean;
}

/**
 * Processed styles which can be passed to `View` elements.
 *
 * @public
 */
export type NativeBlockStyles = TStylesShape['nativeBlockFlow'] &
  TStylesShape['nativeBlockRet'];

/**
 * Processed styles which can be passed to `Text` elements.
 *
 * @public
 */
export type NativeTextStyles = TStylesShape['nativeBlockFlow'] &
  TStylesShape['nativeBlockRet'] &
  TStylesShape['nativeTextFlow'] &
  TStylesShape['nativeTextRet'];

/**
 * Processed Web text styles.
 *
 * @public
 */
export type WebTextStyles = TStylesShape['webTextFlow'];

/**
 * Processed Web block styles.
 *
 * @public
 */
export type WebBlockStyles = TStylesShape['webBlockRet'];

/**
 * Shared properties for all {@link TNode} types.
 *
 * @public
 */
export interface TNodeShape<T extends TNodeType> {
  /**
   * Attributes for this tag.
   */
  readonly attributes: Record<string, string>;
  /**
   * The id for this node, extracted from the id attribute of the
   * underlying {@link Node}.
   */
  readonly id: string | null;
  /**
   * A list of classes for this node, extracted from the class attribute of the
   * underlying {@link Node}.
   */
  readonly classes: string[];
  /**
   * Non-anonymous nodes will hold a reference to a DOM node.
   */
  readonly domNode: Element | null;
  /**
   * The tag name for this node.
   *
   * @remarks Anonymous nodes generated during hoisting won't have a tag name.
   * Also, some TText nodes don't have a tagName.
   */
  readonly tagName: string | null;
  /**
   * The parent of this node before hoisting.
   *
   * @remarks
   * "Before hoisting" implies that this parent will not match "anonymous"
   * parents.
   */
  readonly parent: TDocument | TBlock | TPhrasing | null;
  /**
   * The position of this element relatively to its parents, before hoisting,
   * after collapsing.
   *
   * @remarks
   * "Before hoisting" implies that this index corresponds to the node position
   * in the DOM, after removal of empty elements as per whitespace collapsing
   * algorithm.
   */
  readonly nodeIndex: number;
  /**
   * Children of this node.
   */
  readonly children: ReadonlyArray<TNode>;

  /**
   * Used for debugging purposes.
   */
  readonly displayName: string;

  /**
   * The type of this tnode.
   */
  readonly type: T;

  /**
   * `true` when this tag is not a valid HTML tag **and** there is no custom
   * renderer for this tag.
   */
  readonly isUnregistered: boolean;

  /**
   * Markers form an abstraction in which one node provides semantic information
   * to itself and all its descendants. For example, `ins` elements, which stand
   * for "insertion" of content in the context of an edit will provide the {
   * edits: 'ins' } marker to all its descendants.
   *
   * @remarks This attribute must be considered immutable. Never try to
   * change it by hand, or you might update the markers of an anscestor! For
   * performance reasons, markers instances might be shared from parent to
   * children when they are equal.
   */
  readonly markers: Readonly<Markers>;

  /**
   * TStyles for this TNode. You should not use these unless required.
   * Use {@link TNodeShape.getNativeStyles} instead.
   */
  readonly styles: TStylesShape;

  /**
   * Create a JSX string representation of this node and its children.
   *
   * @remarks The snapshot is _just_ a representation. For example, it will
   * print `href` as an attribute of the TNode, while if you want to access
   * `href` programatically, you'll need to access it via
   * `tnode.attributes.href`.
   *
   * @param options - Customize what should be displayed.
   */
  snapshot(options?: Partial<TNodePrintOptions>): string;

  /**
   * Check if this {@link TNode} has the given `className`.
   *
   * @param className - The class to check.
   */
  hasClass(className: string): boolean;

  /**
   * Test if the given content model matches this TNode content model.
   *
   * @param contentModel - The content model to test against.
   */
  matchContentModel(contentModel: HTMLContentModel): boolean;

  /**
   * Get own native styles.
   */
  getNativeStyles(): T extends 'text' | 'phrasing'
    ? NativeTextStyles
    : NativeBlockStyles;

  /**
   * Get styles that cannot be handled by React Native.
   */
  getWebStyles(): WebTextStyles & WebBlockStyles;
}

/**
 * Metainformation extracted from `<head>` tag.
 *
 * @public
 */
export interface DocumentContext {
  charset: string;
  baseHref: string;
  baseTarget: '_blank' | '_self' | '_parent' | '_top';
  lang: string;
  dir: 'ltr' | 'rtl';
  title: string;
  meta: { name: string; value: string }[];
  links: Record<string, string>[];
}

/**
 * A transient render node (TNode) is a ready-to-render data structure.
 */
export type TNode = TBlock | TDocument | TEmpty | TPhrasing | TText;

/**
 * Transient render nodes to be rendered in React Native `Views`.
 */
export interface TBlock extends TNodeShape<'block'> {
  readonly tagName: string;
  readonly domNode: Element;
}

/**
 * The root of a transient render tree. It has a special
 * `context` field holding metadata about the page.
 */
export interface TDocument extends TNodeShape<'document'> {
  /**
   * An object containing special information for this document, such as lang,
   * dir, charset, baseHref...
   */
  readonly context: Readonly<DocumentContext>;
}

/**
 * Transient render nodes for a non-renderable elements, e.g.
 * for elements which have an {@link HTMLContentModel} set to **none**.
 */
export interface TEmpty extends TNodeShape<'empty'> {
  readonly domNode: Element;
}

/**
 * Transient render nodes to be rendered in React Native `Text`
 * components. TPhrasing nodes have {@link TText}, {@link TPhrasing} or
 * {@link TEmpty} children.
 */
export interface TPhrasing extends TNodeShape<'phrasing'> {}

/**
 * Transient render nodes to be rendered in React Native `Text`
 * components. They don't have children, and have a special `data` field
 * holding the text to be rendered.
 */
export interface TText extends TNodeShape<'text'> {
  readonly data: string;
  readonly textNode: Text;
}

/**
 * Distinct {@link TNode} types.
 */
export type TNodeType = 'block' | 'phrasing' | 'text' | 'empty' | 'document';

export interface TNodeMethods {
  bindChildren(
    this: TNodeImpl,
    children: ReadonlyArray<TNodeImpl>,
    shouldUpdateNodeIndexes?: boolean
  ): void;
  cloneInitParams<T extends TNodeInit = TNodeInit>(
    this: TNodeImpl,
    partial?: Partial<T>
  ): T;
  isCollapsibleLeft(this: TNodeImpl): boolean;
  isCollapsibleRight(this: TNodeImpl): boolean;
  isEmpty(this: TNodeImpl): boolean;
  trimLeft(this: TNodeImpl): void;
  trimRight(this: TNodeImpl): void;
  matchContentModel(contentModel: HTMLContentModel): boolean;
  collapse(this: TNodeImpl): void;
  hasClass(className: string): boolean;
  /**
   * Collpase this node children.
   *
   * @param params
   * @returns A list of indexes to splice
   */
  collapseChildren(this: TNodeImpl): void;
  spliceChildren(this: TNodeImpl, indexes: number[]): void;
  snapshot(options?: Partial<TNodePrintOptions>): string;
  toString(): string;
  initialize(init: TNodeInit): void;
  setMarkers: SetMarkersForTNode;
}

export interface TNodeImpl<T = TNodeInit>
  extends TNodeMethods,
    Omit<TNodeShape<TNodeType>, 'children' | 'parent'> {
  __nodeIndex: number | null;
  __trimmedLeft: boolean;
  __trimmedRight: boolean;
  readonly children: ReadonlyArray<TNodeImpl>;
  readonly init: T;
  readonly parent: TNodeImpl | null;
  readonly hasWhiteSpaceCollapsingEnabled: boolean;
  readonly parentStyles: TStylesShape | null;
  readonly elementModel: HTMLElementModel<string, HTMLContentModel> | null;
  readonly contentModel: HTMLContentModel | null;
}
