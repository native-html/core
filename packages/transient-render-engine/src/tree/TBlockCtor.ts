import HTMLContentModel from '../model/HTMLContentModel';
import TNodeCtor, { GenericTNodeCtor, Mutable } from './TNodeCtor';
import { TNodeImpl, TNodeInit } from './tree-types';

export interface TBlockImpl extends TNodeImpl {}

const TBlockCtor = function TBlock(this: Mutable<TBlockImpl>, init: TNodeInit) {
  this.initialize(init);
} as Function as GenericTNodeCtor<TNodeInit, TBlockImpl>;

//@ts-ignore
TBlockCtor.prototype = new TNodeCtor('block', 'TBlock');

TBlockCtor.prototype.matchContentModel = function matchContentModel(
  contentModel
) {
  return (
    contentModel === HTMLContentModel.block ||
    contentModel === HTMLContentModel.mixed
  );
};

TBlockCtor.prototype.collapseChildren = function collapseChildren() {
  let indexesToSplice: number[] = [];
  this.children.forEach((child, i) => {
    child.collapse();
    if (child.isEmpty()) {
      indexesToSplice.push(i);
    }
  });
  this.spliceChildren(indexesToSplice);
};

TBlockCtor.prototype.getNativeStyles = function getNativeStyles() {
  return {
    ...this.styles.nativeBlockFlow,
    ...this.styles.nativeBlockRet
  };
};

export default TBlockCtor;

export { TBlockCtor };
