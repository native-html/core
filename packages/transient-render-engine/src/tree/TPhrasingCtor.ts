import HTMLContentModel from '../model/HTMLContentModel';
import TNodeCtor, { GenericTNodeCtor, Mutable } from './TNodeCtor';
import { TNodeImpl, TNodeInit } from './tree-types';

export interface TPhrasingImpl extends TNodeImpl {}

function isChildEmpty(c: TNodeImpl) {
  return c.isEmpty();
}

const TPhrasingCtor = function TPhrasing(
  this: Mutable<TNodeImpl>,
  init: TNodeInit
) {
  this.initialize(init);
} as Function as GenericTNodeCtor<TNodeInit, TPhrasingImpl>;

//@ts-ignore
TPhrasingCtor.prototype = new TNodeCtor('phrasing', 'TPhrasing');

TPhrasingCtor.prototype.matchContentModel = function matchContentModel(
  contentModel
) {
  return (
    contentModel === HTMLContentModel.textual ||
    contentModel === HTMLContentModel.mixed
  );
};

TPhrasingCtor.prototype.isEmpty = function isEmpty() {
  // Only anonymous phrasing nodes with every children empty can be considered "empty"
  return this.tagName === null && this.children.every(isChildEmpty);
};

TPhrasingCtor.prototype.collapseChildren = function collapseChildren() {
  let previous: TNodeImpl | null = null;
  this.children.forEach((childK, k) => {
    const j = k - 1;
    childK.collapse();
    if (
      j > -1 &&
      childK.isCollapsibleLeft() &&
      (previous as TNodeImpl).isCollapsibleRight()
    ) {
      // We must trim left from current to support trimming
      // after <br> and <wbr> tags.
      childK.trimLeft();
    }
    previous = childK;
  });
  this.trimLeft();
  this.trimRight();
  return null;
};

export default TPhrasingCtor;

export { TPhrasingCtor };
