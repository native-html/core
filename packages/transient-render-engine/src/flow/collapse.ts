import { TNodeImpl } from '../tree/tree-types';

export function collapse(root: TNodeImpl): TNodeImpl {
  root.collapse();
  root.trimLeft();
  root.trimRight();
  return root;
}
