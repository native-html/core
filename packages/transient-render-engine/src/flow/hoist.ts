import TBlockCtor, { TBlockImpl } from '../tree/TBlockCtor';
import { TNodeImpl } from '../tree/tree-types';
import { TPhrasingCtor } from '../tree/TPhrasingCtor';
import { TTextCtor } from '../tree/TTextCtor';

/**
 * Wrap text nodes around TPhrasing nodes.
 * @param tnode The parent node of all elements to group.
 */
function groupText(tnode: TBlockImpl): TNodeImpl {
  let newChildren: TNodeImpl[] = [];
  const wrapperInit = {
    elementModel: null,
    context: tnode.init.context,
    // We need to merge styles here to make sure
    // some React Native styles working only for the uppermost Text element
    // such as "textAlign" are preserved.
    parentStyles: tnode.styles,
    parent: null
  };
  let wrapper = new TPhrasingCtor(wrapperInit);
  let wrapperChildren: TNodeImpl[] = [];
  for (const child of tnode.children) {
    if (child instanceof TTextCtor || child instanceof TPhrasingCtor) {
      wrapperChildren.push(child);
    } else {
      if (wrapperChildren.length) {
        newChildren.push(wrapper);
        wrapper.bindChildren(wrapperChildren);
        wrapper = new TPhrasingCtor(wrapperInit);
        wrapperChildren = [];
      }
      newChildren.push(child);
    }
  }
  if (wrapperChildren.length) {
    wrapper.bindChildren(wrapperChildren);
    newChildren.push(wrapper);
  }
  tnode.bindChildren(newChildren);
  return tnode;
}

function hoistNode(tnode: TNodeImpl): TNodeImpl {
  if (tnode instanceof TTextCtor) {
    return tnode;
  }
  tnode.bindChildren(tnode.children.map(hoistNode));
  if (tnode instanceof TPhrasingCtor) {
    for (const cnode of tnode.children) {
      if (cnode instanceof TBlockCtor) {
        const newNode = new TBlockCtor(tnode.cloneInitParams());
        newNode.bindChildren(tnode.children);
        const output = groupText(newNode);
        return output;
      }
    }
  } else if (tnode instanceof TBlockCtor) {
    if (tnode.children.length > 0) {
      const output = groupText(tnode);
      return output;
    }
  }
  return tnode;
}

export function hoist(tree: TNodeImpl): TNodeImpl {
  return hoistNode(tree);
}
