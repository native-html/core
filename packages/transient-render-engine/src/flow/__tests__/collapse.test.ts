import { TBlockCtor } from '../../tree/TBlockCtor';
import { TNodeImpl } from '../../tree/tree-types';
import { TPhrasingCtor } from '../../tree/TPhrasingCtor';
import { TTextCtor, TTextImpl } from '../../tree/TTextCtor';
import { collapse } from '../collapse';
import { hoist } from '../hoist';
import {
  deeplyNestedSource,
  nestedHyperlinksSource,
  recursiveHoisting,
  rfc002Source
} from './shared';
import { defaultInit, translateTreeTest } from './utils';
import { Text } from '../../dom/dom-utils';

function makeTTree(
  html: string,
  removeLineBreaksAroundEastAsianDiscardSet = false
): TNodeImpl {
  return collapse(
    hoist(
      translateTreeTest(html, { removeLineBreaksAroundEastAsianDiscardSet })
    )
  );
}

function makeTextChildren() {
  return [new TTextCtor({ textNode: new Text(''), ...defaultInit })];
}

describe('collapse function', () => {
  it('should collapse a tree such as specified in RFC002 example', () => {
    const ttree = makeTTree(rfc002Source);
    expect(ttree).toMatchSnapshot();
  });
  it('should collapse adjacent tags', () => {
    const ttree = makeTTree('<span><span>foo </span><span> bar</span></span>');
    expect(ttree).toMatchSnapshot();
  });
  it('should handle nested anchors', () => {
    const ttree = makeTTree(nestedHyperlinksSource);
    expect(ttree).toMatchSnapshot();
  });
  it('should handle deeply nested HTML', () => {
    const ttree = makeTTree(deeplyNestedSource);
    expect(ttree).toMatchSnapshot();
  });
  it('should hoist blocks recursively', () => {
    const ttree = makeTTree(recursiveHoisting);
    expect(ttree).toMatchSnapshot();
  });
  it('should handle body tag', () => {
    const html = `<body><div>
    <span>Hello world!</span>
</div></body>`;
    const ttree = makeTTree(html);
    expect(ttree).toMatchSnapshot();
  });
  it('should remove empty children from TBlock nodes', () => {
    const ttree = new TBlockCtor(defaultInit);
    ttree.bindChildren(makeTextChildren());
    expect(collapse(ttree).children).toHaveLength(0);
  });
  it('should remove empty anonymous TText children from TPhrasing nodes', () => {
    const ttree = new TPhrasingCtor(defaultInit);
    ttree.bindChildren(makeTextChildren());
    expect(collapse(ttree).children).toHaveLength(0);
  });
  it('should remove empty anonymous TPhrasing children from TPhrasing nodes', () => {
    const ttree = new TPhrasingCtor(defaultInit);
    const tphrasing = new TPhrasingCtor(defaultInit);
    tphrasing.bindChildren(makeTextChildren());
    ttree.bindChildren([tphrasing]);
    const collapsed = collapse(ttree);
    expect(collapsed.children).toHaveLength(0);
    expect(collapsed).toMatchSnapshot();
  });
  it('should remove children from TPhrasing nodes which are empty after timming', () => {
    const ttree = new TPhrasingCtor(defaultInit);
    ttree.bindChildren([
      // This node will be empty after trimming right, and should be removed
      new TTextCtor({ textNode: new Text(' '), ...defaultInit }),
      new TTextCtor({ textNode: new Text(' Foo'), ...defaultInit }),
      new TTextCtor({ textNode: new Text(' Bar'), ...defaultInit })
    ]);
    const collapsed = collapse(ttree);
    expect(collapsed.children).toHaveLength(2);
    expect(collapsed).toMatchSnapshot();
  });
  it('should handle direct style inheritance', () => {
    const ttree = makeTTree(
      '<div style="font-size: 18px;border-width: 20px;"><span style="color: red;">This is nice!</span></div>'
    );
    expect(ttree).toMatchSnapshot();
  });
  it('should handle indirect style inheritance', () => {
    const ttree = makeTTree(
      '<div style="font-size: 18px;"><div style="border-width: 20px;">This is great!</div></div>'
    );
    expect(ttree).toMatchSnapshot();
  });
  it('should not collapse when white-space CSS property is set to "pre"', () => {
    const ttree = makeTTree(
      '<div style="white-space: pre;">  This is great!  </div>'
    );
    expect(ttree).toMatchSnapshot();
  });
  it('should collapse when white-space CSS property is set to "normal"', () => {
    const ttree = makeTTree(
      '<div style="white-space: normal;">  This is great!  </div>'
    );
    expect(ttree).toMatchSnapshot();
  });
  it('should collapse when white-space CSS property is not set', () => {
    const ttree = makeTTree('<div>  This is great!  </div>');
    expect(ttree).toMatchSnapshot();
  });
  it('should collapse a node with white-space set to "normal" while its parent has white-space set to "pre"', () => {
    const ttree = makeTTree(
      '<div style="white-space: pre;"><span> This is nice </span><strong style="white-space: normal"> Should collapse </strong></div>'
    );
    // left space of <strong> child DOMTextNode should be spared, since left
    // sibling is not collapsible (tested in Mozilla Firefox and Chromium)
    expect(ttree).toMatchSnapshot();
  });
  it('should withold TEmpty nodes', () => {
    const ttree = makeTTree(
      '<div><span>Hi!</span><link rel="author" href="mailto:don@company.com" /></div>'
    );
    expect(ttree).toMatchSnapshot();
  });
  it('should support removeLineBreaksAroundEastAsianDiscardSet param', () => {
    const span = makeTTree('<span>\u2F00\n\u2FDA</span>', true) as TTextImpl;
    expect(span).toMatchSnapshot();
    expect(span.data).toBe('\u2F00\u2FDA');
  });
  it('should set `nodeIndex` field corresponding to the actual index relative to parent', () => {
    const src = `<table>
    <tr>
      <th>Month</th>
      <th>Fortnight</th>
      <th>Savings</th>
    </tr>
    <tr>
      <td>January</td>
      <td>first</td>
      <td>$50</td>
    </tr>
    <tr>
      <td>second</td>
      <td>$80</td>
      <td>$80</td>
    </tr>
  </table>`;
    const ttree = makeTTree(src);
    expect(ttree.tagName).toBe('table');
    ttree.children.forEach((child, i) => {
      expect(child.nodeIndex).toBe(i);
    });
  });
  it('should handle <br> tags', () => {
    const src = `<p>
    should<br>
    collapse
    </p>`;
    const ttree = makeTTree(src);
    expect(ttree).toMatchSnapshot();
  });
  it('should retain whitespaces between block and textual nodes', () => {
    const src = '<div><strong>Tags:</strong>\n<a href="">hello</a>';
    const ttree = makeTTree(src);
    expect(ttree).toMatchSnapshot();
  });
});
