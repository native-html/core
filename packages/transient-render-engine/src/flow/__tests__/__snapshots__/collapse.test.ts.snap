// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`collapse function should collapse a node with white-space set to "normal" while its parent has white-space set to "pre" 1`] = `
<TBlock tagName="div" nodeIndex={0} webStyles={{ whiteSpace: "pre"}}>
  <TPhrasing anonymous nodeIndex={0} webStyles={{ whiteSpace: "pre"}}>
    <TText tagName="span" nodeIndex={0} data=" This is nice " webStyles={{ whiteSpace: "pre"}} />
    <TText tagName="strong" nodeIndex={1} data=" Should collapse" webStyles={{ whiteSpace: "normal"}} />
  </TPhrasing>
</TBlock>
`;

exports[`collapse function should collapse a tree such as specified in RFC002 example 1`] = `
<TBlock tagName="a" nodeIndex={0} href="https://domain.com">
  <TPhrasing anonymous nodeIndex={0}>
    <TText anonymous nodeIndex={0} data="This is " />
    <TText tagName="span" nodeIndex={1} data="phrasing content" />
  </TPhrasing>
  <TBlock tagName="img" nodeIndex={1} src="https://domain.com/logo.jpg" />
  <TPhrasing anonymous nodeIndex={2}>
    <TText anonymous nodeIndex={0} data="and this is " />
    <TText tagName="strong" nodeIndex={1} data="too" />
    <TText anonymous nodeIndex={2} data="." />
  </TPhrasing>
</TBlock>
`;

exports[`collapse function should collapse adjacent tags 1`] = `
<TPhrasing tagName="span" nodeIndex={0}>
  <TText tagName="span" nodeIndex={0} data="foo " />
  <TText tagName="span" nodeIndex={1} data="bar" />
</TPhrasing>
`;

exports[`collapse function should collapse when white-space CSS property is not set 1`] = `
<TBlock tagName="div" nodeIndex={0}>
  <TPhrasing anonymous nodeIndex={0}>
    <TText anonymous nodeIndex={0} data="This is great!" />
  </TPhrasing>
</TBlock>
`;

exports[`collapse function should collapse when white-space CSS property is set to "normal" 1`] = `
<TBlock tagName="div" nodeIndex={0} webStyles={{ whiteSpace: "normal"}}>
  <TPhrasing anonymous nodeIndex={0} webStyles={{ whiteSpace: "normal"}}>
    <TText anonymous nodeIndex={0} data="This is great!" webStyles={{ whiteSpace: "normal"}} />
  </TPhrasing>
</TBlock>
`;

exports[`collapse function should handle <br> tags 1`] = `
<TBlock tagName="p" nodeIndex={0}>
  <TPhrasing anonymous nodeIndex={0}>
    <TText anonymous nodeIndex={0} data="should" />
    <TText tagName="br" nodeIndex={1} data="" />
    <TText anonymous nodeIndex={2} data="collapse" />
  </TPhrasing>
</TBlock>
`;

exports[`collapse function should handle body tag 1`] = `
<TBlock tagName="body" nodeIndex={0}>
  <TBlock tagName="div" nodeIndex={0}>
    <TPhrasing anonymous nodeIndex={0}>
      <TText tagName="span" nodeIndex={0} data="Hello world!" />
    </TPhrasing>
  </TBlock>
</TBlock>
`;

exports[`collapse function should handle deeply nested HTML 1`] = `
<TBlock tagName="article" nodeIndex={0}>
  <TPhrasing anonymous nodeIndex={0}>
    <TText anonymous nodeIndex={0} data="First implicit paragraph." />
  </TPhrasing>
  <TBlock tagName="div" nodeIndex={1}>
    <TPhrasing anonymous nodeIndex={0}>
      <TText anonymous nodeIndex={0} data="Second implicit paragraph" />
    </TPhrasing>
    <TBlock tagName="p" nodeIndex={1}>
      <TPhrasing anonymous nodeIndex={0}>
        <TText anonymous nodeIndex={0} data="A new paragraph. " />
        <TText tagName="span" nodeIndex={1} data="And a span within" />
      </TPhrasing>
      <TBlock tagName="img" nodeIndex={1} src="https://domain.com/logo.jpg" />
    </TBlock>
  </TBlock>
</TBlock>
`;

exports[`collapse function should handle direct style inheritance 1`] = `
<TBlock tagName="div" nodeIndex={0} nativeStyles={{ borderTopWidth: 20, borderRightWidth: 20, borderBottomWidth: 20, borderLeftWidth: 20 }}>
  <TPhrasing anonymous nodeIndex={0} nativeStyles={{ fontSize: 18 }}>
    <TText tagName="span" nodeIndex={0} data="This is nice!" nativeStyles={{ fontSize: 18, color: "red" }} />
  </TPhrasing>
</TBlock>
`;

exports[`collapse function should handle indirect style inheritance 1`] = `
<TBlock tagName="div" nodeIndex={0}>
  <TBlock tagName="div" nodeIndex={0} nativeStyles={{ borderTopWidth: 20, borderRightWidth: 20, borderBottomWidth: 20, borderLeftWidth: 20 }}>
    <TPhrasing anonymous nodeIndex={0} nativeStyles={{ fontSize: 18 }}>
      <TText anonymous nodeIndex={0} data="This is great!" nativeStyles={{ fontSize: 18 }} />
    </TPhrasing>
  </TBlock>
</TBlock>
`;

exports[`collapse function should handle nested anchors 1`] = `
<TPhrasing tagName="a" nodeIndex={0} href="https://domain.com">
  <TText anonymous nodeIndex={0} data="This is " />
  <TText tagName="span" nodeIndex={1} data="phrasing content" />
  <TText anonymous nodeIndex={2} data=" " />
  <TPhrasing tagName="a" nodeIndex={3} href="https://sub.domain.com">
    <TText anonymous nodeIndex={0} data="and this is " />
    <TText tagName="strong" nodeIndex={1} data="too" />
  </TPhrasing>
</TPhrasing>
`;

exports[`collapse function should hoist blocks recursively 1`] = `
<TBlock tagName="span" nodeIndex={0}>
  <TPhrasing anonymous nodeIndex={0}>
    <TText anonymous nodeIndex={0} data="Line1" />
  </TPhrasing>
  <TBlock tagName="strong" nodeIndex={1}>
    <TPhrasing anonymous nodeIndex={0}>
      <TText anonymous nodeIndex={0} data="Line2 " />
      <TText tagName="span" nodeIndex={1} data="Line3" />
    </TPhrasing>
    <TBlock tagName="img" nodeIndex={1} src="https://domain.com/logo.jpg" />
  </TBlock>
</TBlock>
`;

exports[`collapse function should not collapse when white-space CSS property is set to "pre" 1`] = `
<TBlock tagName="div" nodeIndex={0} webStyles={{ whiteSpace: "pre"}}>
  <TPhrasing anonymous nodeIndex={0} webStyles={{ whiteSpace: "pre"}}>
    <TText anonymous nodeIndex={0} data="  This is great!  " webStyles={{ whiteSpace: "pre"}} />
  </TPhrasing>
</TBlock>
`;

exports[`collapse function should remove children from TPhrasing nodes which are empty after timming 1`] = `
<TPhrasing anonymous nodeIndex={0}>
  <TText anonymous nodeIndex={0} data="Foo" />
  <TText anonymous nodeIndex={1} data=" Bar" />
</TPhrasing>
`;

exports[`collapse function should remove empty anonymous TPhrasing children from TPhrasing nodes 1`] = `<TPhrasing anonymous nodeIndex={0} />`;

exports[`collapse function should retain whitespaces between block and textual nodes 1`] = `
<TBlock tagName="div" nodeIndex={0}>
  <TPhrasing anonymous nodeIndex={0}>
    <TText tagName="strong" nodeIndex={0} data="Tags:" />
    <TText anonymous nodeIndex={1} data=" " />
    <TText tagName="a" nodeIndex={2} data="hello" href="" />
  </TPhrasing>
</TBlock>
`;

exports[`collapse function should support removeLineBreaksAroundEastAsianDiscardSet param 1`] = `<TText tagName="span" nodeIndex={0} data="⼀⿚" />`;

exports[`collapse function should withold TEmpty nodes 1`] = `
<TBlock tagName="div" nodeIndex={0}>
  <TPhrasing anonymous nodeIndex={0}>
    <TText tagName="span" nodeIndex={0} data="Hi!" />
  </TPhrasing>
  <TEmpty tagName="link" nodeIndex={1} href="mailto:don@company.com" />
</TBlock>
`;
